Sources/NIOHTTPCompression/HTTPResponseDecompressor.swift

```diff
@@ -0,0 +1,197 @@
+        switch request {
+        case .head(var head):
+            if head.headers.contains(name: "Accept-Encoding") {
+                context.write(data, promise: promise)
+            } else {
+                head.headers.replaceOrAdd(name: "Accept-Encoding", value: "deflate, gzip")
+                context.write(self.wrapOutboundOut(.head(head)), promise: promise)
+            }
+        default:
+            context.write(data, promise: promise)
+        }
+    }
+
+    public func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+        switch self.unwrapInboundIn(data) {
+        case .head(let head):
+            let algorithm: CompressionAlgorithm?
+            let contentType = head.headers[canonicalForm: "Content-Encoding"].first?.lowercased()
+            if contentType == "gzip" {
+                algorithm = .gzip
+            } else if contentType == "deflate" {
+                algorithm = .deflate
+            } else {
+                algorithm = nil
+            }
+
+            let length = head.headers[canonicalForm: "Content-Length"].first.flatMap { Int($0) }
+
+            if let algorithm = algorithm, let length = length {
+                do {
+                    try self.initializeDecoder(encoding: algorithm, length: length)
+                } catch {
+                    context.fireErrorCaught(error)
+                }
+            }
+
+            context.fireChannelRead(data)
+        case .body(var part):
+            switch self.state {
+            case .compressed(_, let originalLength):
+                while part.readableBytes > 0 {
+                    do {
+                        var buffer = context.channel.allocator.buffer(capacity: 16384)
+                        try self.stream.inflatePart(input: &part, output: &buffer)
+                        self.inflated += buffer.readableBytes
+
+                        if self.limit.exceeded(compressed: originalLength, decompressed: self.inflated) {
+                            context.fireErrorCaught(DecompressionError.limit)
+                            return
+                        }
+
+                        context.fireChannelRead(self.wrapInboundOut(.body(buffer)))
+                    } catch {
+                        context.fireErrorCaught(error)
+                        return
+                    }
+                }
+            default:
+                context.fireChannelRead(data)
+            }
+        case .end:
+            deflateEnd(&self.stream)
+            context.fireChannelRead(data)
+        }
+    }
+
+    private func initializeDecoder(encoding: CompressionAlgorithm, length: Int) throws {
+        self.state = .compressed(encoding, length)
+
+        self.stream.zalloc = nil
+        self.stream.zfree = nil
+        self.stream.opaque = nil
+
+        let window: Int32
+        switch encoding {
+        case .gzip:
+            window = 15 + 16
+        default:
+            window = 15
+        }
```

Comment:
```
var window: Int32 = 15
if case .gzip = encoding {
  window += 16
}
```
more readable?

Reply:
How about moving it to a property in `CompressionAlgorithm`?